File: src/codes/Users_Manual/advanced_examples/custom_modulation_0.rs
# extern crate autd3;
# extern crate tokio;
# extern crate autd3_driver;
use autd3::prelude::*;
use autd3_driver::derive::*;

#[derive(Modulation)]
pub struct Burst {
    config: SamplingConfiguration,
    loop_behavior: LoopBehavior,
}

impl Burst {
    pub fn new() -> Self {
        Self { 
            config: SamplingConfiguration::from_frequency(4e3).unwrap(),
            loop_behavior: LoopBehavior::Infinite,
        }
    }
}

impl Modulation for Burst {
    fn calc(&self) -> Result<Vec<EmitIntensity>, AUTDInternalError> {
        Ok((0..4000)
            .map(|i| if i == 3999 { EmitIntensity::MAX } else { EmitIntensity::MIN })
            .collect())
    }
}
# fn main() { 
# }

File: src/codes/Users_Manual/advanced_examples/custom_modulation_0.py
import numpy as np
from pyautd3 import EmitIntensity, SamplingConfig
from pyautd3.modulation import Modulation


class Burst(Modulation):
    _length: int

    def __init__(self, length: int = 4000):
        super().__init__(SamplingConfig.freq(4000))
        self._length = length

    def calc(self):
        buf = np.array([EmitIntensity.minimum()] * self._length)
        buf[0] = EmitIntensity.maximum()
        return buf
